<h1>Solution for Data Collection</h1>

Task 1

How many CYP protein structures are available in ChEMBL?

1. To solve this problem, it is necessary to extract all targets that <b>contain</b> the word <b>Cytochrome</b> in their names
2. The task does not specify which type of organism we are interested. So let's calculate how many CYP structures are available for each type of organism

In [1]:
import pandas as pd
from chembl_webresource_client.new_client import new_client

In [2]:
# gether protein chembl id of interested target from ChEMBL
def download_targetid_chembl(taget_name):
    target = new_client.target
    
    # instead of looking for exact protein name in ChEMBL we are looking for proteins which contain  the word
    cyp = target.filter(pref_name__icontains=taget_name).only(
    ['target_chembl_id', 'target_organism', 'target_pref_name'])
    
    return cyp

In [3]:
res = download_targetid_chembl('Cytochrome')

In [4]:
# Converting data into pandas dataframe format
cyp = pd.DataFrame(data=res)

In [5]:
# Let's take a look at the data
cyp.head()

organism  \
0                                         Homo sapiens   
1    Candida albicans (strain SC5314 / ATCC MYA-287...   
2                                         Homo sapiens   
3                                         Homo sapiens   
4                                         Homo sapiens   
..                                                 ...   
122                                  Rattus norvegicus   
123                                       Homo sapiens   
124                                       Homo sapiens   
125                                       Homo sapiens   
126                                Bacillus megaterium   

                                             pref_name target_chembl_id  
0                                 Cytochrome P450 11A1       CHEMBL2033  
1                                   Cytochrome P450 51       CHEMBL1780  
2                                 Cytochrome P450 19A1       CHEMBL1978  
3                                 Cytochrome P450 11B1       CHEMBL1908  
4                                  Cytochrome P450 7A1       CHEMBL1851  
..                                                 ...              ...  
122                     NADH-cytochrome b5 reductase 3    CHEMBL4523194  
123                                Cytochrome P450 4F8    CHEMBL4523270  
124                                Cytochrome P450 4Z1    CHEMBL4523375  
125                                    Cytochrome P450    CHEMBL4523986  
126  Bifunctional cytochrome P450/NADPH--P450 reduc...    CHEMBL4630872  

[127 rows x 3 columns]

In [9]:
print(len(cyp), 'CYP protein structures in ChEMBL')

127 CYP protein structures in ChEMBL


In [7]:
# Let's estimate how many different types of organisms are presented
cyp['organism'].describe()

count              127
unique              22
top       Homo sapiens
freq                53
Name: organism, dtype: object

The results tell us that there are 127 lines in total, of which 22 are unique. The most common string in the column is <i>Homo sapiens</i>, it occurs 53 times

In [13]:
# Print all CYP structures which belongs to Homo sapiens
cyp[cyp['organism'] == 'Homo sapiens'].sort_values('pref_name')

organism                                          pref_name  \
112  Homo sapiens  Apoptotic protease-activating factor 1/Caspase...   
125  Homo sapiens                                    Cytochrome P450   
0    Homo sapiens                               Cytochrome P450 11A1   
3    Homo sapiens                               Cytochrome P450 11B1   
23   Homo sapiens                               Cytochrome P450 11B2   
35   Homo sapiens                               Cytochrome P450 17A1   
2    Homo sapiens                               Cytochrome P450 19A1   
108  Homo sapiens                                 Cytochrome P450 1A   
17   Homo sapiens                                Cytochrome P450 1A1   
30   Homo sapiens                                Cytochrome P450 1A2   
6    Homo sapiens                                Cytochrome P450 1B1   
24   Homo sapiens                                 Cytochrome P450 21   
52   Homo sapiens                               Cytochrome P450 24A1   
51   Homo sapiens                               Cytochrome P450 26A1   
111  Homo sapiens                               Cytochrome P450 26B1   
106  Homo sapiens                               Cytochrome P450 2A13   
42   Homo sapiens                                Cytochrome P450 2A6   
43   Homo sapiens                                Cytochrome P450 2B6   
18   Homo sapiens                               Cytochrome P450 2C18   
34   Homo sapiens                               Cytochrome P450 2C19   
7    Homo sapiens                                Cytochrome P450 2C8   
33   Homo sapiens                                Cytochrome P450 2C9   
21   Homo sapiens                                Cytochrome P450 2D6   
41   Homo sapiens                                Cytochrome P450 2E1   
54   Homo sapiens                                Cytochrome P450 2J2   
95   Homo sapiens                                Cytochrome P450 2W1   
88   Homo sapiens                                 Cytochrome P450 3A   
39   Homo sapiens                                Cytochrome P450 3A4   
84   Homo sapiens                            Cytochrome P450 3A4/3A5   
64   Homo sapiens                               Cytochrome P450 3A43   
31   Homo sapiens                                Cytochrome P450 3A5   
96   Homo sapiens                                Cytochrome P450 3A7   
19   Homo sapiens                               Cytochrome P450 4A11   
117  Homo sapiens                               Cytochrome P450 4F11   
100  Homo sapiens                               Cytochrome P450 4F12   
13   Homo sapiens                                Cytochrome P450 4F2   
123  Homo sapiens                                Cytochrome P450 4F8   
124  Homo sapiens                                Cytochrome P450 4Z1   
20   Homo sapiens                                 Cytochrome P450 51   
4    Homo sapiens                                Cytochrome P450 7A1   
66   Homo sapiens  Cytochrome P450, family 4, subfamily X, polype...   
70   Homo sapiens                       Cytochrome b-245 heavy chain   
72   Homo sapiens                  Cytochrome b-c1 complex subunit 7   
59   Homo sapiens                                      Cytochrome b5   
121  Homo sapiens                               Cytochrome b5 type B   
86   Homo sapiens                                       Cytochrome c   
61   Homo sapiens                     Cytochrome c oxidase subunit 1   
62   Homo sapiens                     Cytochrome c oxidase subunit 2   
115  Homo sapiens         Cytochrome c1, heme protein, mitochondrial   
22   Homo sapiens                       NADH-cytochrome b5 reductase   
85   Homo sapiens                   NADPH--cytochrome P450 reductase   
107  Homo sapiens                       Putative cytochrome P450 2D7   
9    Homo sapiens  Ubiquinol-cytochrome C reductase complex 7.2 k...   

    target_chembl_id  
112    CHEMBL3885517  
125    CHEMBL4523986  
0         CHEMBL2033  
3         CHEMBL1908  
23   